contracts/exchangeIssuance/FlashMintWrapped.sol

```diff
@@ -17,220 +17,1087 @@
+    );
+  }
+
+  /**
+   * Swaps and then wraps each _requiredComponents sequentially based on _swapData and _wrapData
+   *
+   * @param _inputToken                 Input token that will be sold
+   * @param _maxAmountInputToken        Maximum amount of input token that can be spent
+   * @param _swapData                   ComponentSwapData (input token -> underlyingERC20) for each _requiredComponents in the same order
+   * @param _invokeWrapData             ComponentInvokeWrapData (underlyingERC20 -> wrapped Set component) for each _requiredComponents in the same order
+   * @param _requiredComponents         Issuance components gotten from DebtIssuanceModule.getRequiredComponentIssuanceUnits()
+   * @param _requiredAmounts            Issuance units gotten from DebtIssuanceModule.getRequiredComponentIssuanceUnits()
+   */
+  function _swapAndWrapComponents(
+    IERC20 _inputToken,
+    uint256 _maxAmountInputToken,
+    ComponentSwapData[] calldata _swapData,
+    ComponentInvokeWrapData[] calldata _invokeWrapData,
+    address[] memory _requiredComponents,
+    uint256[] memory _requiredAmounts
+  ) internal {
+    // if the required set components contain the input token, we have to make sure that the required amount
+    // for issuance is actually still left over at the end of swapping and wrapping
+    uint256 requiredLeftOverInputTokenAmount = 0;
+
+    // for each component in the swapData / wrapData / requiredComponents array:
+    // 1. swap from input token to unwrapped component (exact to buyUnderlyingAmount)
+    // 2. wrap from unwrapped component to wrapped component (unless unwrapped component == wrapped component)
+    // 3. ensure amount in contract covers required amount for issuance
+    for (uint256 i = 0; i < _requiredComponents.length; ++i) {
+      // if the required set component is the input token, no swapping or wrapping is needed
+      uint256 requiredAmount = _requiredAmounts[i];
+      if (address(_inputToken) == _requiredComponents[i]) {
+        requiredLeftOverInputTokenAmount = requiredAmount;
+        continue;
+      }
+
+      // snapshot balance of required component before swap and wrap operations
+      uint256 componentBalanceBefore = IERC20(_requiredComponents[i]).balanceOf(address(this));
+
+      // swap input token to underlying token
+      _swapToExact(
+        _inputToken, // input
+        IERC20(_swapData[i].underlyingERC20), // output
+        _swapData[i].buyUnderlyingAmount, // buy amount: must come from flash mint caller, we do not know the exchange rate wrapped <> unwrapped
+        _maxAmountInputToken, // maximum spend amount: _maxAmountInputToken as transferred by the flash mint caller
+        _swapData[i].dexData // dex path fees data etc.
+      );
+
+      uint256 underlyingBalance = IERC20(_swapData[i].underlyingERC20).balanceOf(address(this));
+      require(underlyingBalance >= _swapData[i].buyUnderlyingAmount, "FlashMint: UNDERBOUGHT_COMPONENT");
```

Comment:
This doesn't account for any underlying balance that might have been left over in the contract no ?

Reply:
Note that this underlyingBalance is only used here to fail early. 
But you are right, even for that it would have to take into account the previously present balance. 

I wasn't sure if it is worth the gas here though, especially when fixing this... It would just fail either at wrapping or at componentAmountObtained...

Would be very helpful to get your opinion on that

Reply:
I removed this in the newest version, I don't think it adds much value as is, it would require another snapshot which I think is not worth it since it will fail in the next step anyway